filebeat/input/v2/input.go

```diff
@@ -83,25 +88,202 @@ type Context struct {
+		metricsLog.Errorf(
+			"failed to generate uuid to track input metrics register/unregister: %v",
+			err)
+	} else {
+		uid = uidv4.String()
+	}
+
+	// Log the registration to ease tracking down duplicate ID registrations.
+	// Logged at INFO rather than DEBUG since it is not in a hot path and having
+	// the information available by default can short-circuit requests for debug
+	// logs during support interactions.
+	metricsLog.Infow("registering",
+		"input_type", inputType,
+		"id", id,
+		"registry_name", metricsID,
+		"uuid", uid)
+
+	unreg := func() {
+		metricsLog.Infow("unregistering",
+			"input_type", inputType,
+			"id", id,
+			"registry_name",
+			metricsID, "uuid", uid)
+		parentRegistry.Remove(metricsID)
+		// it's safe to make this call even if registering them failed.
+		inputmon.UnregisterMetrics(metricsID)
+	}
+	return Context{
+		ID:            id,
+		IDWithoutName: idWithoutName,
+		Name:          inputType,
+
+		Agent:          agent,
+		Cancelation:    cancelation,
+		StatusReporter: statusReporter,
+
+		Logger: log,
+
+		monitoringRegistry:       reg,
+		monitoringRegistryCancel: unreg,
+	}
+}
+
+func (c *Context) UpdateStatus(status status.Status, msg string) {
 	if c.StatusReporter != nil {
 		c.Logger.Debugf("updating status, status: '%s', message: '%s'", status.String(), msg)
 		c.StatusReporter.UpdateStatus(status, msg)
 	}
 }
 
+// MetricRegistry returns the metrics registry associated with this context.
+// This should be the metrics registry used by inputs to register their metrics.
+// It's already registered to be published by the HTTP monitoring endpoint.
+// If the context wasn't created by NewContext and its monitoring registry is
+// nil, a new registry is created and returned.
+func (c *Context) MetricRegistry() *monitoring.Registry {
+	// It's a precaution in case the context wasn't created by NewContext.
+	if c.monitoringRegistry == nil {
+		c.monitoringRegistry = monitoring.NewRegistry()
+	}
+
+	return c.monitoringRegistry
+}
+
+// UpdateMetricRegistry overrides the `id` and `input` entries in the metrics
+// registry and returns the registry. It exists for backwards compatibility as
+// some inputs set a different input type/name than their name.
+func (c *Context) UpdateMetricRegistry(id, inputType string) *monitoring.Registry {
+	monitoring.NewString(c.MetricRegistry(), "input").Set(inputType)
+	monitoring.NewString(c.MetricRegistry(), "id").Set(id)
+
+	return c.MetricRegistry()
+}
+
+// UnregisterMetrics removes the metrics registry from its parent registry and
+// from the HTTP monitoring endpoint.
+func (c *Context) UnregisterMetrics() {
+	if c.monitoringRegistryCancel != nil {
+		c.monitoringRegistryCancel()
+	}
+}
+
+// PipelineClientListener returns the PipelineClientListener for this context.
+// It collects pipeline metrics for this input on the metrics registry
+// associated with this context.
+func (c *Context) PipelineClientListener() *PipelineClientListener {
```

Comment:
(Repeating from offline thread) Please don't propagate the listener this way. The discussed design was meant to avoid inputs having to correctly manage their own boilerplate. This adds a new call that every input is required to make and pass through when connecting to the pipeline, and on top of that requires them to correctly compose it with any existing client listener callbacks they already had. The client listener should be managed by the API as previously suggested; a straightforward way to do that is to wrap the pipeline with `pipetool.WithClientConfigEdit` to compose the needed listeners with any user-provided configuration.